libyul/ASTNodeRegistry.cpp

```diff
@@ -0,0 +1,150 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libyul/ASTNodeRegistry.h>
+
+#include <libyul/Exceptions.h>
+
+#include <fmt/format.h>
+
+#include <range/v3/algorithm/max.hpp>
+#include <range/v3/view/map.hpp>
+
+using namespace solidity::yul;
+
+ASTNodeRegistry::ASTNodeRegistry(): m_labels{"", ghostPlaceholder}, m_idToLabelMapping{0, 1} {}
+
+ASTNodeRegistry::ASTNodeRegistry(std::vector<std::string> const& _labels, std::vector<size_t> const& _idToLabelMapping)
+{
+	yulAssert(_labels.size() >= 2);
+	yulAssert(_labels[0].empty());
+	yulAssert(_labels[1] == ghostPlaceholder);
+	yulAssert(_idToLabelMapping.size() >= 2);
+	yulAssert(_idToLabelMapping[0] == 0);
+	yulAssert(_idToLabelMapping[1] == 1);
+	std::vector<uint8_t> labelVisited (_labels.size(), false);
```

Comment:
Why `uint8_t`? You seem to be storing only boolean values to it.
```suggestion
	std::vector<bool> labelVisited(_labels.size(), false);
```

Is it a performance microoptimization due to `vector<bool>` being specialized to prioritize space-efficiency? If so, it needs a comment.

Reply:
Also, I see you're not checking that there are no unused labels. I guess it's not strictly necessary for correct behavior, but if we're already iterating over IDs, it would not be much more effort to add counter for unique IDs and assert at the end that it matches the number of labels minus 2.

Reply:
> Is it a performance microoptimization due to `vector<bool>` being specialized to prioritize space-efficiency? If so, it needs a comment.

Added such a comment, that was indeed the rationale.

> Also, I see you're not checking that there are no unused labels. I guess it's not strictly necessary for correct behavior, but if we're already iterating over IDs, it would not be much more effort to add a counter for unique IDs and assert at the end that it matches the number of labels minus 2.

True, added!